File: backend-java/02-backend-java-prototype/src/main/java/com/ganatan/modules/profession/ProfessionService.java
package com.ganatan.modules.profession;

import java.util.List;
import java.util.Optional;

public class ProfessionService {

    private final ProfessionRepository repository;

    public ProfessionService(ProfessionRepository repository) {
        this.repository = repository;
    }

    public List<Profession> getItems() {
        return repository.getItems();
    }

    public Optional<Profession> getItemById(int id) {
        return repository.getItemById(id);
    }

    public Profession createItem(Profession profession) {
        return repository.createItem(profession);
    }

    public Optional<Profession> updateItem(int id, Profession updatedData) {
        return repository.updateItem(id, updatedData);
    }

    public boolean deleteItem(int id) {
        return repository.deleteItem(id);
    }
}

File: backend-javascript/03-backend-javascript-postgresql/src/modules/profession/profession.service.js
import { HTTP_STATUS } from '../../shared/constants/http.js';

const MESSAGES = {
  ITEM_ALREADY_EXISTS: 'Profession already exists',
};

import { validateItem } from './profession.schema.js';

class Service {
  constructor(repository) {
    this.repository = repository;
  }

  async getItems(query) {
    return await this.repository.getItems(query);
  }

  async getItemsCount() {
    return await this.repository.getItemsCount();
  }

  async getItemsPaginated({ page, size, name }) {
    const [items, total] = await Promise.all([
      this.repository.getItems({ page, size, name }),
      this.repository.getItemsCount(name),
    ]);

    return { items, total };
  }

  async getItemById(id) {
    return await this.repository.getItemById(id);
  }

  async createItem(createdData) {
    try {
      validateItem(createdData);
    } catch (error) {
      error.status = HTTP_STATUS.BAD_REQUEST;
      throw error;
    }

    const exists = await this.repository.existsByName(createdData.name);
    if (exists) {
      const error = new Error(MESSAGES.ITEM_ALREADY_EXISTS);
      error.status = HTTP_STATUS.CONFLICT;
      throw error;
    }

    return await this.repository.createItem(createdData);
  }

  async updateItem(id, updatedData) {
    try {
      validateItem(updatedData);
    } catch (error) {
      error.status = HTTP_STATUS.BAD_REQUEST;
      throw error;
    }

    return await this.repository.updateItem(id, updatedData);
  }

  async deleteItem(id) {
    return await this.repository.deleteItem(id);
  }
}

export default Service;
